General fixes for LPC1768

p3p opened the pull request:
ADC filters to work around really bad temperature reading
Initial USB mass storage lock to stop hosts accessing the sd card at the same time
Missing serial class methods
PlatformIO Re-ARM debug build fix
Simplified delayMicroseconds, if it was interrupted at the wrong time it could have caused issues.

Also Added a timeout in setup() for the serial connection.

p3p committed:
fixed some include paths

p3p committed:
LPC1768: add filters to ADC

This is to try and compensate for hardware issue and oversensitivity to noise

p3p committed:
LPC1768: remove the polling section of delayMicroseconds

p3p committed:
LPC1768: lock usb mass storage device while device accesses it.

Currently only applicable to persistent store,
The device always has priority and will unmount the sd card from the host, Windows then tries to automount again so it can look like the explorer window freezes. Linux Mint, by default, just closes the Nemo window.

p3p committed:
Add timeout to make sure if Serial never connects that Marlin still boots

p3p committed:
Remove unneeded ifdef CPU_32_BIT

In general the need for ifdef CPU_32_BIT blocks means that something is missing from the HAL API or a Platform, in this case HAL_TICKS_PER_US was missing from the AVR Platform

p3p committed:
LPC1768: relocate RE-ARM debug_extra_script.py

p3p commented:
I rebased this on the current bugfix-2.0.x before creating the PR but have just realised that bugfix-2.0.x doesn't boot for me atm (watchdog reset), I'l have to look into that tomorrow.

thinkyhead commented:
If it's a general issue, hopefully we'll get a patch for it soon. I've been porting patches from 1.1.x over to 2.0.x, and while I am being extra cautious, small mistakes are always possible. I've got to run off for a few hours to work on … TMC5130 support (w00t!) … but will be scrubbing through PRs and Issues later today.

thinkyhead commented on Marlin/src/HAL/HAL_LPC1768/HAL.cpp:
```diff
@@ -121,14 +122,52 @@ void HAL_adc_start_conversion(const uint8_t adc_pin) {
+uint16_t lowpass_filter(uint16_t value) {
+  const uint8_t k_data_shift = ADC_LOWPASS_K_VALUE;
+  static uint32_t data_delay[NUM_ANALOG_INPUTS] = { 0 };
+  uint32_t &active_filter = data_delay[active_adc];
```
Does it matter if `active_adc` isn't set yet? I see that it defaults to 0, and will most likely get set by `HAL_adc_start_conversion`, but if a bad index is given then it would remain set to 0.

Obviously very unlikely and probably safe. But, just thought I'd point it out. Is there any state where no ADC conversion might be busy? As I recall, for the most part once Marlin boots up it starts doing ADC and never stops.

p3p replied:
The worst that could happen, if there is a mistake and someone doesn't call `HAL_adc_start_conversion` before `HAL_adc_get_result`, is ADC 0 will be used, presumably returning an invalid reading as it wasn't started (index 0 will always be valid at least) .

thinkyhead commented on Marlin/src/HAL/HAL_LPC1768/HAL.cpp:
```diff
@@ -121,14 +122,52 @@ void HAL_adc_start_conversion(const uint8_t adc_pin) {
+#endif
+#ifdef ADC_USE_LOWPASS_FILTER
+  data = lowpass_filter((uint16_t)data);
+#endif
```
Insert obligatory comment about indentation here.

p3p replied:
Insert obligatory comment about it looking better this way.

( I'l fix it ;) )

thinkyhead commented on frameworks/CMSIS/LPC1768/lib/usb/mscuser.h:
```diff
@@ -49,21 +49,56 @@ enum class Sense_KEY : uint8_t {
+  Sense_KEY key;
+  Sense_ASC asc;
+  Sense_ASCQ ascq;
+};
```
If there will only ever be one `Sense` object, this class can be made all-static to use function-like calling semantics, which are slightly smaller and faster. I realize this is far less important on 32-bit platforms, but it would be consistent. That said, this class does look like more than one instance could be used.

p3p replied:
You're right this class could be static its still very much a work in progress, I'm still hoping I can find a nice compatible, fully functional usb stack to replace all this with, I'm not familiar enough with the intricacies (although a lot more now than I was) and workarounds for OS differences that are needed. (if only Windows used the Lock/Unlock functionality built into the damn protocol)

thinkyhead commented on frameworks/CMSIS/LPC1768/lib/usb/mscuser.h:
```diff
@@ -38,7 +38,7 @@ extern void MSC_SetCSW (void);
 extern void MSC_BulkOut(void);
 
-enum class Sense_KEY : uint8_t {
+enum struct Sense_KEY : uint8_t {
```
Hmm! This is a new C++ feature to me. I like it.

p3p commented:
Found the lockup issue, the LPC1768 serial class didn't implement the return -1 on an empty receive buffer, returned 0 instead, the xon/xoff commit changes the logic in `get_serial_commands` to need it.

p3p committed:
LPC1768: Fix Serial API

Add missing serial methods used if TX_BUFFER_SIZE is set
Change return value of HalSerial:read to match Arduino API

p3p commented:
@thinkyhead Updated with changes plus added the fix for LPC1768 locking up after boot.

Roxy-3D commented:
@p3p   Are we ready to merge now?   It would be nice to get all these changes into BugFix-2.0.0 !

thinkyhead commented:
Moiged!
